packages/api/src/WithServerEvents.tsx

```diff
@@ -0,0 +1,241 @@
+    key: string,
+    listener: (event: IMessageEvent) => void
+  ) => void;
+  unregisterChangeListener: (key: string) => void;
+  unregisterMessageListener: (key: string) => void;
+}
+
+const RECONNECT_TIME = 5000;
+
+export class WithServerEvents extends React.Component<
+  IWithEventsProps,
+  IWithEventsState
+> {
+  private starting = false;
+  private started = false;
+  private retries = 0;
+  private preferredProtocol: string = '';
+  private eventSource: EventSource | undefined;
+  private webSocket: WebSocket | undefined;
+  private apiEndpoint: string | undefined;
+  private headers: IHeaders | undefined;
+  private changeListeners = {};
+  private messageListeners = {};
+
+  public constructor(props: IWithEventsProps) {
+    super(props);
+    this.state = {
+      registerChangeListener: this.registerChangeListener.bind(this),
+      registerMessageListener: this.registerChangeListener.bind(this),
+      unregisterChangeListener: this.unregisterChangeListener.bind(this),
+      unregisterMessageListener: this.unregisterMessageListener.bind(this),
+    };
+  }
+
+  public registerChangeListener(
+    key: string,
+    listener: (event: IChangeEvent) => void
+  ) {
+    this.changeListeners[key] = listener;
+  }
+
+  public registerMessageListener(
+    key: string,
+    listener: (event: IMessageEvent) => void
+  ) {
+    this.messageListeners[key] = listener;
+  }
+
+  public unregisterChangeListener(key: string) {
+    delete this.changeListeners[key];
+  }
+
+  public unregisterMessageListener(key: string) {
+    delete this.messageListeners[key];
+  }
+
+  public async componentDidMount() {
+    this.apiEndpoint = this.props.apiUri;
+    this.headers = this.props.headers;
+    this.start();
+  }
+  public render() {
+    return this.props.children(this.state);
+  }
+
+  private close() {
+    this.started = false;
+    this.starting = false;
+    if (this.webSocket) {
+      this.webSocket.close();
+    }
+    if (this.eventSource) {
+      this.eventSource.close();
+    }
+  }
+
+  private start() {
+    if (this.starting || this.started) {
+      return;
+    }
+    this.startConnection(this.retries % 2 === 0);
+  }
+
+  private onFailure(error: any) {
+    this.retries++;
+    // Initialy retry very quickly.
+    let reconnectIn = RECONNECT_TIME;
+    if (this.retries < 3) {
+      reconnectIn = 1;
+    }
+    setTimeout(() => {
+      this.close();
+      console.log('Reconnecting');
+      switch (this.preferredProtocol) {
+        // Once we find a protocol that works, keep using it.
+        case 'ws':
+          this.startConnection(true);
+          break;
+        case 'es':
+          this.startConnection(false);
+          break;
+        default:
+          // Keep flipping between WS and ES untill we find one that works.
+          this.startConnection(this.retries % 2 === 0);
+          break;
+      }
+    }, reconnectIn);
+  }
+
+  private async startConnection(connectUsingWebSockets: boolean) {
```

Comment:
If you don't like writing a bunch of `this.props.apiUri` you can set a `const {apiUri: apiEndpoint} = this.props;` at the top of the function.

But I found this to lessen the IDE refactoring capabilities when I actually want to rename the prop because the name I picked sucks, since I end up getting something like `const { newCoolName: oldBadName} = this.props`.